srcts/src/components/_shinyResizeObserver.ts

```diff
@@ -0,0 +1,152 @@
+/**
+ * A resize observer that ensures Shiny outputs resize during or just after
+ * their parent container size changes. Useful, in particular, for sidebar
+ * transitions or for full-screen card transitions.
+ *
+ * @class ShinyResizeObserver
+ * @typedef {ShinyResizeObserver}
+ */
+class ShinyResizeObserver {
+  /**
+   * The actual ResizeObserver instance.
+   * @private
+   * @type {ResizeObserver}
+   */
+  private resizeObserver: ResizeObserver;
+  /**
+   * An array of elements that are currently being watched by the Resize
+   * Observer.
+   *
+   * @details
+   * We don't currently have lifecycle hooks that allow us to unobserve elements
+   * when they are removed from the DOM. As a result, we need to manually check
+   * that the elements we're watching still exist in the DOM. This array keeps
+   * track of the elements we're watching so that we can check them later.
+   * @private
+   * @type {HTMLElement[]}
+   */
+  private resizeObserverEntries: HTMLElement[];
+
+  /**
+   * Watch containers for size changes and ensure that Shiny outputs and
+   * htmlwidgets within resize appropriately.
+   *
+   * @details
+   * The ShinyResizeObserver is used to watch the containers, such as Sidebars
+   * and Cards for size changes, in particular when the sidebar state is toggled
+   * or the card body is expanded full screen. It performs two primary tasks:
+   *
+   * 1. Dispatches a `resize` event on the window object. This is necessary to
+   *    ensure that Shiny outputs resize appropriately. In general, the window
+   *    resizing is throttled and the output update occurs when the transition
+   *    is complete.
+   * 2. If an output with a resize method on the output binding is detected, we
+   *    directly call the `.onResize()` method of the binding. This ensures that
+   *    htmlwidgets transition smoothly. In static mode, htmlwidgets does this
+   *    already.
+   *
+   * @note
+   * This resize observer also handles race conditions in some complex
+   * fill-based layouts with multiple outputs (e.g., plotly), where shiny
+   * initializes with the correct sizing, but in-between the 1st and last
+   * renderValue(), the size of the output containers can change, meaning every
+   * output but the 1st gets initialized with the wrong size during their
+   * renderValue(). Then, after the render phase, shiny won't know trigger a
+   * resize since all the widgets will return to their original size (and thus,
+   * Shiny thinks there isn't any resizing to do). The resize observer works
+   * around this by ensuring that the output is resized whenever its container
+   * size changes.
+   * @constructor
+   */
+  constructor() {
+    this.resizeObserverEntries = [];
+    this.resizeObserver = new ResizeObserver((entries) => {
+      const resizeEvent = new Event("resize");
```

Comment:
I vaguely recall that there are some scenarios where this "modern" version doesn't actually work (I think because of the way we currently listen to resize events), which is why the [card resizing logic](https://github.com/rstudio/bslib/blob/main/R/card.R#L299-L300) does it this way:

```suggestion
      var resizeEvent = window.document.createEvent('UIEvents');
      resizeEvent.initUIEvent('resize', true, false, window, 0);
```

Reply:
I originally used that pattern anticipating that there were reasons, but TypeScript warns that [`.initUIEvent()` is deprecated](https://developer.mozilla.org/en-US/docs/Web/API/UIEvent/initUIEvent). From what I can tell, this pattern was used to [satisfy Internet Explorer](https://stackoverflow.com/a/1818513).

Reply:
Ok, as long as the current approach works without shiny (on modern browsers), then lets use it.